### PR TITLE
Feat: Pin component

### PR DESCRIPTION
**Pull request by khanghoang**

### CODE REVIEW

**Comment by gaearon** [comment]
Can you provide a gif or vide of how it works?

**Comment by khanghoang** [comment]
@gaearon here it is 

pin

**Comment by FezVrasta** [comment]
Does it work between page reloads?

**Comment by gaearon** [comment]
Not currently but if we implement something like https://github.com/facebook/react-devtools/issues/295 then it should be easy to make it work for pinning too. Would you like to try to tackle https://github.com/facebook/react-devtools/issues/295?

**Comment by FezVrasta** [comment]
I haven't free time for this but it would be awesome to have this PR and #295 work together, this PR alone doesn't look much interesting to me because I would have to redo everything on each reload 😓

**Comment by gaearon** [comment]
This is a good start. Can you make the UI in the panel prettier?

**Comment by khanghoang** [comment]
@gaearon which part that you want me to change and how will it look like?

**Comment by gaearon** [comment]
I think having two identical panels is a bit confusing. I would expect the pinned panel to not have its own footer, I think. I would also expect some sort of header explaining what it is, or another way to separate them.

**Comment by khanghoang** [comment]
@gaearon I made some changes, please take a look if they're what you're looking for :)

container example 2017-02-26 13-35-25

**Comment by gaearon** [comment]
Can we leave the checkmarks at the top bar?

**Comment by khanghoang** [comment]
@gaearon like this?

container example 2017-02-28 11-15-44

If it's not what you want then please let me know what it's gonna look like.

**Comment by khanghoang** [comment]
ping @gaearon

**Comment by FezVrasta** [comment]
I think he means to have the "Trace React Updates", "Highlight Search" etc on the top bar.

**Comment by khanghoang** [comment]
@FezVrasta you're right 👍

**Comment by khanghoang** [comment]
@gaearon @FezVrasta here it is

container example 2017-03-04 00-05-48

**Comment by FezVrasta** [comment]
Just my 2cents, I'm not part of the project, but that "Pinned Components" header is huge...

**Comment by gaearon** [comment]
Can you make it styled more blandly? Perhaps similar to how "No search results" is styled right now, but a little bit larger. Also, it is weird that the main tree is centered. I would expect that after pinned components, there is a horizontal line, and then the normal trees go after them as normal, without centering.

**Comment by khanghoang** [comment]
@gaearon sorry for the late reply. How is about this?
container example 2017-03-16 21-29-23

**Comment by khanghoang** [comment]
@gaearon ping

**Comment by khanghoang** [comment]
@gaearon can you help me review this one?

**Comment by gaearon** [comment]
I'll review it—as I mentioned in other threads, I'm in the middle of a big React change and will jump on other tasks and repos after that's done.

**Comment by khanghoang** [comment]
@gaearon this is just a friendly reminder if you can take a look at this PR when you have time.

**Comment by gaearon** [comment]
I will look at it—as you can see I’ve been merging more urgent bugfixes, and there are still more bugs that I need to fix before cutting a release. This is more experimental stuff, and while I’m excited you worked on this, it’s less urgent and I can only look at it when I have some free time (currently I’m jumping between several large task lists in several projects).

It would be really helpful if you could look at the persistence of the collapsed/expanded tree state in the meantime separately, as that’s a pre-requisite to making this feature truly useful.

**Comment by khanghoang** [comment]
hi @gaearon, I finally figure out a way to persist pinned components, I think we can use the same technique for collapsed/expanded tree state. Take a look at this PR if you have time.

File: frontend/Store.js [hunk 1]
```diff
@@ -128,6 +130,7 @@ class Store extends EventEmitter {
     this._parents = new Map();
     this._nodesByName = new Map();
     this._bridge = bridge;
+    this.pinnedComponents = JSON.parse(localStorage.getItem('foo', JSON.stringify(this.pinnedComponents))) || [];
```
Comment (petetnt) [inline frontend/Store.js]: Should be `JSON.parse(localStorage.getItem('foo')) || []`

File: frontend/Store.js [hunk 2]
```diff
@@ -564,6 +567,50 @@ class Store extends EventEmitter {
...
+    ];
+    localStorage.setItem('foo', JSON.stringify(this.pinnedComponents));
+    this.emit('unpinComponent');
+  }
+
+
+  getPath(componentId: string): string {
+    const currentNode = this.get(componentId);
+    if (!currentNode) return null;
+    const key = currentNode.get('key') === null ? '' : currentNode.get('key');
```
Comment (petetnt) [inline frontend/Store.js]: `currentNode.get('key') || ''` might be clearer.

File: frontend/Store.js [hunk 3]
```diff
@@ -564,6 +567,50 @@ class Store extends EventEmitter {
...
+    let path = `${currentNode.get('name')}${key}`;
+    let temp = this.getParent(componentId);
+    while (this.getParent(temp)) {
+      const parentNode = this.get(this.getParent(temp));
+      const k = parentNode.get('key') === null ? '' : parentNode.get('key');
```
Comment (petetnt) [inline frontend/Store.js]: `parentNode.get('key') || ''` might be clearer.

File: frontend/SearchPane.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  reload: PropTypes.func,
+  searchText: PropTypes.string,
+  selectFirstSearchResult: PropTypes.func,
+  onChangeSearch: PropTypes.func,
+  placeholderText: PropTypes.string,
+};
+
+var Wrapped = decorate({
+  listeners(props) {
+    return ['searchText', 'placeholderchange'];
```
Comment (petetnt) [inline frontend/SearchPane.js]: What is placeholderchange? 🤔

File: frontend/Store.js (hunk 1 above)
Comment (petetnt) [reply frontend/Store.js]: The key should probably be something else than `foo` 😸

File: frontend/Store.js [hunk 5]
```diff
@@ -103,6 +103,8 @@ class Store extends EventEmitter {
   placeholderText: string;
   preferencesPanelShown: boolean;
   refreshSearch: boolean;
+  // pinned components
```
Comment (petetnt) [inline frontend/Store.js]: Comment not needed

**Comment by bvaughn** [comment]
React DevTools has been rewritten and [recently launched a new version 4 UI](https://reactjs.org/blog/2019/08/15/new-react-devtools.html) 🎉 The source code for this rewrite was done in a separate repository and is [being merged into the main React repo](https://github.com/facebook/react/pull/16381).

**Thank you for taking the time to contribute to this extension** 🙇‍♂️

Because this PR is for the version code base, I am closing it. If you'd like to contribute to the new extension, please visit [the new repository](https://github.com/facebook/react/issues?q=is%3Aissue+is%3Aopen+label%3A%22Component%3A+Developer+Tools%22).

**Pull request closed without merging**